cpp/include/rapidsmpf/buffer/pinned_memory_resource.hpp

```diff
@@ -0,0 +1,406 @@
+ * (page-locked) host memory asynchronously using CUDA streams.
+ */
+class PinnedMemoryResource {
+  public:
+    /**
+     * @brief Friend function to get the host_accessible property.
+     */
+    friend constexpr void get_property(
+        const PinnedMemoryResource&, cuda::mr::host_accessible
+    ) noexcept {}
+
+    /**
+     * @brief Constructs a new pinned memory resource.
+     *
+     * @param pool The pinned memory pool to use for allocations.
+     *
+     * @throws rapidsmpf::cuda_error If the pinned memory resource is not supported for
+     * the current CUDA version.
+     */
+    PinnedMemoryResource(PinnedMemoryPool& pool);
+
+    /**
+     * @brief Destroys the pinned memory resource.
+     *
+     * Note: This does not deallocate memory that was allocated through this resource.
+     * All allocated memory should be explicitly deallocated before destruction.
+     */
+    ~PinnedMemoryResource();
+
+    /**
+     * @brief Allocates pinned memory asynchronously.
+     *
+     * @param stream The CUDA stream to use for the allocation operation.
+     * @param bytes The number of bytes to allocate.
+     * @return A pointer to the allocated memory, or nullptr if allocation failed.
+     */
+    void* allocate(rmm::cuda_stream_view stream, size_t bytes);
+
+    /**
+     * @brief Allocates pinned memory asynchronously with alignment.
+     *
+     * @param stream The CUDA stream to use for the allocation operation.
+     * @param bytes The number of bytes to allocate.
+     * @param alignment The alignment requirement for the allocation.
+     * @return A pointer to the allocated memory, or nullptr if allocation failed.
+     */
+    void* allocate(rmm::cuda_stream_view stream, size_t bytes, size_t alignment);
+
+    /**
+     * @brief Allocates pinned memory synchronously.
+     *
+     * @param bytes The number of bytes to allocate.
+     * @param alignment The alignment requirement for the allocation.
+     * @return A pointer to the allocated memory, or nullptr if allocation failed.
+     */
+    void* allocate_sync(size_t bytes, size_t alignment);
+
+    /**
+     * @brief Deallocates pinned memory asynchronously.
+     *
+     * @param stream The CUDA stream to use for the deallocation operation.
+     * @param ptr A pointer to the memory to deallocate.
+     * @param bytes The size of the memory to deallocate.
+     */
+    void deallocate(rmm::cuda_stream_view stream, void* ptr, size_t bytes) noexcept;
+
+    /**
+     * @brief Deallocates pinned memory asynchronously with alignment.
+     *
+     * @param stream The CUDA stream to use for the deallocation operation.
+     * @param ptr A pointer to the memory to deallocate.
+     * @param bytes The size of the memory to deallocate.
+     * @param alignment The alignment that was used for allocation.
+     */
+    void deallocate(
+        rmm::cuda_stream_view stream, void* ptr, size_t bytes, size_t alignment
+    ) noexcept;
```

Comment:
How do you even use alignment for deallocation? Isn't the pointer all you need/should use?